atmat/atphysics/Radiation/FDW.c

```diff
@@ -0,0 +1,465 @@
+{	/* Find Ohmi's diffusion matrix BDIFF (la integrada) of a wiggler
+	   BDIFF - cumulative Ohmi's diffusion is initialized to 0
+	   BDIFF is preallocated 1 dimensional array to store 6-by-6 matrix 
+	   columnwise.
+	*/
+	
+	int m;	
+	double  MKICK[36], BKICK[36];
+
+	/* 4-th order symplectic integrator constants */
+	int Niter = Nstep*(le/Lw);
+    double SL = Lw/Nstep;
+	double dl1 = SL*KICK1;
+    double dl0 = SL*KICK2;
+	
+	/* Allocate memory for thin kick matrix MKICK
+	   and a diffusion matrix BKICK
+	*/
+	for(m=0; m < 6; m++)
+		{	MKICK[m] = 0;
+			BKICK[m] = 0;
+		}
+	
+	double zEndPointH[2];
+    double zEndPointV[2];
+	double ax,ay,axpy,aypx;
+	double B[3];
+	double E0G;
+	double factorvect;
+    struct gwigR pWig;
+    int flag;       
+    
+	/* Transform orbit to a local coordinate system of an element
+       BDIFF stays zero	*/
+    if(T1) ATaddvv(orbit_in,T1);	
+    if(R1) ATmultmv(orbit_in,R1);	
+
+    /*Generate the wiggler*/ 
+    zEndPointH[0] = 0;
+    zEndPointH[1] = le;
+    zEndPointV[0] = 0;
+    zEndPointV[1] = le;
+	
+	/* Energy is defined in the lattice in eV but GeV is used by the gwig code. */
+	E0G = E0 / 1e9; 
+    GWigInit2(&pWig,E0G,le,Lw,Bmax,Nstep,Nmeth,NHharm,NVharm,0,0,zEndPointH,zEndPointV,By,Bx,T1,T2,R1,R2);
+    
+    /* Propagate orbit_in and BDIFF through a 4-th order integrator */
+    flag=1;
+    GWigGauge(&pWig, orbit_in, flag);
+ 
+    GWigAx(&pWig, orbit_in, &ax, &axpy);
+    GWigAy(&pWig, orbit_in, &ay, &aypx);
+    GWigB(&pWig, orbit_in, B);
+	factorvect = vectorialnB2(&pWig, orbit_in, B);
+    orbit_in[1] -= ax;
+    orbit_in[3] -= ay;
+	
+    GWigRadiationKicks(&pWig, orbit_in, factorvect, SL);
+    orbit_in[1] += ax;
+    orbit_in[3] += ay;
+	
+    for (m = 1; m <= Niter; m++ )   /* Loop over slices	*/
+		{	   
+			wigglerM(&pWig, orbit_in, dl1, MKICK);
+			wigglerB(&pWig, orbit_in, dl1, BKICK);
+			ATsandwichmmt(MKICK,BKICK);
+		    ATaddmm(BKICK,BDIFF);
+	   	    GWigMap_2nd(&pWig, orbit_in, dl1);
+                
+			wigglerM(&pWig, orbit_in, dl0, MKICK);
+			wigglerB(&pWig, orbit_in, dl0, BKICK);
+			ATsandwichmmt(MKICK,BKICK);
+			ATaddmm(BKICK,BDIFF);
+            GWigMap_2nd(&pWig, orbit_in, dl0);
+               
+			wigglerM(&pWig, orbit_in, dl1, MKICK);
+			wigglerB(&pWig, orbit_in, dl1, BKICK);
+			ATsandwichmmt(MKICK,BKICK);
+			ATaddmm(BKICK,BDIFF);
+            GWigMap_2nd(&pWig, orbit_in, dl1);
+			   
+			GWigAx(&pWig, orbit_in, &ax, &axpy);
+			GWigAy(&pWig, orbit_in, &ay, &aypx);
+            GWigB(&pWig, orbit_in, B);
+			factorvect = vectorialnB2(&pWig, orbit_in, B);
+			orbit_in[1] -= ax;
+			orbit_in[3] -= ay;
+			
+			GWigRadiationKicks(&pWig, orbit_in, factorvect, SL);
+			orbit_in[1] += ax;
+			orbit_in[3] += ay;
+		}      
+        
+    flag=-1;
+    GWigGauge(&pWig, orbit_in, flag);
+		
+    if(R2) { 
+	   ATmultmv(orbit_in,R2);
+	   ATsandwichmmt(R2,BDIFF);
+	}
+    if(T2) {
+		ATaddvv(orbit_in,T2);	
+		ATsandwichmmt(T2,BDIFF);
+	}

```

Comment:
T2 is not a 6x6 transfer matrix, it's a vector ! You cannot send it to `ATsandwichmmt`. Did you try and check the result ? It should crash…

A translation has no effect on beam matrices. There is nothing to do to BDIFF.

Reply:
That is right. It is now corrected.